Modify old image_ocr.py example to work with tf.keras (Tensorflow 2.2.0)

joshuacwnewton opened the pull request:
# Purpose
This pull request fixes issues that were preventing the old `image_ocr.py` example from working in Tensorflow 2.2.0.

These fixes are likely small enough to not warrant a pull request, but I was hoping to get feedback at least once on my style/formatting at the start. I have mostly used Git in solo projects, so I wanted to check if there were any bad habits I had developed in my solo work.

_(Side note: I typo'd my email address in my local Git config and realized halfway through, which is why it shows as two different commit authors here.)_

# Changes
- Include alternate requirements.txt file specific to the image_ocr,py example to make setup quicker
- Update import statements to work with `tf.keras`
- Bugfix for `TextImageGenerator` initialization steps that were called too late
- Bugfix for visualization string formatting (list of strings -> ndarray)

# Request for feedback
- [ ] General commit/pull request style critiques and advice

Commit:
Add new requirements.txt for image_ocr.py imports

The old image_ocr.py example requires additional imports to function.
To keep the existing requirements.txt file untouched, this commit adds
a second file to make setting up a virtual environment easier.

Commit:
Modify Keras imports to support tf.keras

The old image_ocr.py example was written for Multi-Backend Keras. This
commit provides compatibility for the latest tf.keras module.

joshuacwnewton committed:
Convert 'source_str' from list of str to ndarray

Input data batches from TextImageGenerator contain `source_str`, which
is used inside the visualization callback. It is a list of strings that
correspond to the word represented by each image. Tensorflow throws an
error because it expects a Tensor, but `source_str` can't be
converted to a Tensor. Using np.array(`source_str`) fixes this.

joshuacwnewton committed:
Move on_train_begin() steps to __init__()

TextImageGenerator has some initialization steps that occur in
on_train_begin(). This method is called at the beginning of model.fit().
However, this is too late for these steps. An error occurs earlier,
because img_gen.next_train() is called before model.fit().

flosincapite commented on examples/vision/image_ocr.py:
```diff
@@ -211,6 +212,12 @@ def __init__(self, monogram_file, bigram_file, minibatch_size,
         self.blank_label = self.get_output_size() - 1
         self.absolute_max_string_len = absolute_max_string_len
 
+        # Previously included in self.on_train_begin
```
Why the switch?

joshuacwnewton replied:
Another method, `next_train()`, relies on the results of these steps. But, `self.on_train_begin()` is only called at the beginning of `model.fit()` ([see explanation here.](https://www.tensorflow.org/guide/keras/custom_callback#an_overview_of_callback_methods)) So, [this line](https://github.com/MLH-Fellowship/keras-io/blob/f2d7aa8b8201eefc56cf6218b3ed16c0b847e125/examples/vision/image_ocr.py#L549) fails, because `next_train()` is called before `model.fit()` is called.

I've clarified the wording in the comment in another commit.

flosincapite replied:
So I would do two things:

1) pull that other commit with the updated comment in to this PR (since it's relevant);
2) leave `on_train_begin()` intact but just make sure it's called once before any call to next_train is made.

~

One way to accomplish 2):

in `__init__`, put some kind of sentinel variable, e.g.

```
  [...]
  self._wordlist_was_built = False
  [...]
```

Then, in `on_train_begin`, set that variable to `True`. Then, in `next_train`, you can check the state of `_wordlist_was_built` and, if it's `False`, call `on_train_begin`.

~

The motivation here is to ensure that nothing unexpected happens. With your implementation (`on_train_begin` runs in `__init__`), if somebody were to use `model.fit` instead of `model.fit_generator`, `on_train_begin` would run twice (once in `__init__` and once when `model.fit` explicitly calls it). That could cause unexpected consequences.

joshuacwnewton replied:
For 1, is there something else I need to to do pull the commits into this PR? Right now, I'm seeing them already underneath this discussion. ("joshuacwnewton added 3 commits...") But, I may be misunderstanding how PRs work.

For 2, I like what you've proposed. It preserves much more of the original implementation than my changes did. Thank you for making such a detailed suggestion. :)

joshuacwnewton replied:
Aha, I can link commits by their hash! That should make referencing them easier.

I've implemented your suggestions for 2 in commit 9e97d65. :)

flosincapite commented on requirements-ocr.txt:
```diff
@@ -0,0 +1,12 @@
+pygments~=2.4.2
```
A lot of these requirements are duplicated from the top-level requirements.txt.

If this example has extra reqs, just place them in the examples/ directory (sibling to the code).

joshuacwnewton replied:
Thanks! That makes sense. I've made those changes. :)

joshuacwnewton committed:
Remove redundant ocr requirements and add comment

joshuacwnewton committed:
Move ocr requirements.txt to examples/vision

joshuacwnewton committed:
Clarify comment about moved on_train_begin() steps

joshuacwnewton committed:
Use flag in next_train() to call on_train_begin()

next_train() is dependent on initialization handled by the
on_train_begin() callback method. If next_train() is called before the
callback is invoked, then next_train() will fail. This commit adds a
flag to keep track of whether the callback has executed, and manually
invokes the callback if it hasn't already been called.

parthivapsani reviewed (approved):
Your `self._train_was_initialized` flag looks good to me, and it seems like it would prevent multiple initialization calls
